cirq/value/linear_dict.py

```diff
@@ -0,0 +1,263 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Linear combination represented as mapping of things to coefficients."""
+
+from typing import (Any, Callable, Dict, ItemsView, Iterable, Iterator,
+                    KeysView, Mapping, overload, Tuple, TypeVar, Union,
+                    ValuesView)
+
+Scalar = Union[complex, float]
+TVector = TypeVar('TVector')
+
+_TDefault = TypeVar('_TDefault')
+
+
+class LinearDict(Dict[TVector, Scalar]):
+    """Represents linear combination of things.
+
+    LinearDict implements the basic linear algebraic operations of vector
+    addition and scalar multiplication for linear combinations of abstract
+    vectors. Keys represent the vectors, values represent their coefficients.
+    The only requirement on the keys is that they be hashable.
```

Comment:
Do they also have to support equality (think you delegate to vector equality in __eq__?

Reply:
Yes! This is implied by ['hashable'](https://docs.python.org/3/glossary.html#term-hashable). Added clarification.